include/seqan3/io/alignment_file/detail.hpp

```diff
@@ -227,4 +231,189 @@ std::string get_cigar_string(alignment_type && alignment,
+ *          soft clipped bases at the start of the alignment and (5) the number of bases at the end of the alignment.
+ *
+ * \details
+ *
+ * For example, the view over the cigar string "1S4M1D2M2S" will return `{[(M,4), (D,1), (M,2)], 7, 6, 1, 2}`.
+ */
+template <std::ranges::InputRange cigar_input_type>
+std::tuple<std::vector<std::pair<char, size_t>>, size_t, size_t, size_t, size_t>
+parse_cigar(cigar_input_type && cigar_input)
+{
+    auto constexpr is_cigar_op = is_char<'M'> || is_char<'I'> || is_char<'D'> || is_char<'S'> ||
+                                 is_char<'H'> || is_char<'='> || is_char<'X'> || is_char<'P'> || is_char<'N'>;
+    std::vector<std::pair<char, size_t>> operations{};
+    size_t sc_begin_count{}; // number of soft clipped bases at the beginning
+    size_t sc_end_count{};   // number of soft clipped bases at the end
+    std::array<char, 20> buffer{}; // buffer to parse numbers with from_chars. Biggest number should fit in uint64_t
+    char cigar_op{'\0'};
+    size_t cigar_count{0};
+    size_t ref_length{0}, seq_length{0}; // length of aligned part for ref and query
+
+    auto update_lengths_fn = [&ref_length, &seq_length, &cigar_op, &cigar_count] ()
+        {
+            if (is_char<'M'>(cigar_op) || is_char<'='>(cigar_op) || is_char<'X'>(cigar_op))
+            {
+                ref_length += cigar_count;
+                seq_length += cigar_count;
+            }
+            else if (is_char<'D'>(cigar_op) || is_char<'N'>(cigar_op))
+            {
+                ref_length += cigar_count;
+            }
+            else if (is_char<'I'>(cigar_op))
+            {
+                seq_length += cigar_count;
+            }
+            else // illegal character
+            {
+                if (is_char<'P'>(cigar_op))
+                    throw parse_error{"We do currently not support cigar operation 'P'."};
+                else
+                    throw parse_error{std::string{"Illegal cigar operation: "} + std::string{1, cigar_op}};
+            }
+        };
+
+    // transform input into a signle input view if it isn't already
+    auto cigar_view = cigar_input | view::single_pass_input;
+
+    // check hard/soft clipping at the beginning manually
+    // -----------------------------------------------------------------------------------------------------------------
+    auto [ignore, buffer_end] = std::ranges::copy(cigar_view | view::take_until_or_throw(is_cigar_op), buffer.data());
+    (void) ignore;
+
+    cigar_op = *std::ranges::begin(cigar_view);
+    std::ranges::next(std::ranges::begin(cigar_view));
+
+    if (is_char<'H'>(cigar_op)) // hard clipping is ignored. parse the next operation
+    {
+        auto [ignore2, buffer_end2] = std::ranges::copy(cigar_view
+                                                        | view::take_until_or_throw(is_cigar_op), buffer.data());
```

Comment:
I am wondering if we did not decide on putting the operator before the new line?

Reply:
The syntax we agreed on for chaining views was:

```cpp
auto v = foo
       | view::one
       | view::two;
```

i.e. aligningthe views on top of each other and the pipe symbols also and the pipe symbols with the assignment operator (if present)

Reply:
Boah, did we? Didn't we realise how ugly this is and absolutely not consistent with breaking after an operator?

Reply:
Yes, but it's nice to read, because it clearly communicates that every line adds another view, you don't ahve to scan to the end of the line. And it follows the examples from range-v3:
https://ericniebler.github.io/range-v3/

Reply:
Great, so we have the opportunity to make it better 👍 
Like in all other places we need to do it consistent. But like for every operator it is so much easier to read until the end of the line and then see an operator at the end which directly indicates there is more coming on the next line (because if you read the expression you will automatically be at the end with your eyes and then, oh boy, there it is the operator suggesting there is more coming or the semicolon. So we already expect that there is something at the end but now there is only void. Nothing but the rest of the unbounded screen and your eyes travel to the edge of the monitor getting lost from the actual content. 

```cpp
auto v = much_better |
         view::one |
         view::two_with_longer_name |
         view::oh_this_is_really_looking_good |
         view::told_ya;
```